### PR TITLE
Basic integration with the electrum server electrs

### PR DESCRIPTION
**Pull request by dagurval**
This is a MVP for integrating electrum server with Bitcoin Unlimited.

To enable the server, you need to run `make electrs` and add `electrum=1` to `bitcoin.conf`. Adding `debug=electrum` will enable more useful logging.

### CODE REVIEW

**Comment by dagurval** [comment]
A way to test this with Electron Cash is by running 

    ./electron-cash -v --oneserver --server=127.0.0.1:60001:t --testnet

and creating a wallet with the tpub key `tpubD6NzVbkrYhZ4YRDtddy5nLmvdH7Qn6oR7euDpjqnZdniaJDTbaL17Gq86bsVNhKMkYwGvSvhamz5QkouzGJ4e2rkyHWbF5mHGX5Up377zBM`

Electron Cash also has added a nice feature to rebuild the wallet history `Wallet -> Rebuild history`.

Beware of this bug in electron cash https://github.com/Electron-Cash/Electron-Cash/issues/1222

**Review by sickpig — COMMENTED** [review]
@dagurval

What about adding some words about what elecrtrs is?

File: src/electrum/electrs.cpp [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+{
+    // look for electrs in same path as bitcoind
+    const std::string def =
+        boost::process::search_path("electrs", {boost::dll::program_location().remove_filename()}).string();
+
+    const std::string path = GetArg("-electrumexec", def);
+
+    if (!boost::filesystem::exists(path))
+    {
+        std::stringstream ss;
```
Comment (gandrewstone) [inline src/electrum/electrs.cpp:31]: To help the end user, let's have a different error message if the path is "".  Maybe "Electrum Server electrs was not found alongside bitcoind.  Please specify a location using -electrumexec"

**Comment by gandrewstone** [comment]
I did an out-of-source-tree build but the electrs binary still goes to src/electrs.  To replicate this, from your base git checkout dir:
./autogen.sh
mkdir debug
cd debug
../configure [optional: --enable-debug --enable-shared]
make V=1
make electrs

File: src/electrum/electrumserver.cpp [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    if (!Logging::LogAcceptCategory(ELECTRUM))
+    {
+        return;
+    }
+
+    std::stringstream ss;
+    ss << path;
+    for (auto &a : args)
+    {
+        ss << a << " ";
```
Comment (gandrewstone) [inline src/electrum/electrumserver.cpp]: ss << " " << a;

Or you get an executable named "electrs-vvvv", and an extra space at the end I presume, (in the log only).

**Comment by gandrewstone** [comment]
I get:
```
2019-04-03 19:47:16 Electrum: spawning /fast/bitcoin/buelectrum/src/electrs -vvvv --daemon-rpc-addr=localhost:18332 --daemon-dir=/home/stone/.bitcoin/testnet3 --jsonrpc-import --db-dir=/home/stone/.bitcoin/testnet3/electrs --network=testnet --monitoring-addr=127.0.0.1:4224 --cookie=thezerg:34983429q980dkdkdsakd --txid-limit=500
2019-04-03 19:47:16 Electrum: stderr log thread started.
2019-04-03 19:47:16 Electrum: stdout log thread started.
2019-04-03 19:47:16 Electrum: server exited with exit code: 32
2019-04-03 19:47:16 Error: Unable to start RPC services. See debug log for details.
```
BUT when I run the spawn line directly on the command line I get:
```
/fast/bitcoin/buelectrum/src/electrs -vvvv --daemon-rpc-addr=localhost:18332 --daemon-dir=/home/stone/.bitcoin/testnet3 --jsonrpc-import --db-dir=/home/stone/.bitcoin/testnet3/electrs --network=testnet --monitoring-addr=127.0.0.1:4224 --cookie=thezerg:34983429q980dkdkdsakd --txid-limit=500
Config { log: StdErrLog { verbosity: Trace, quiet: false, timestamp: Off, modules: [], writer: "stderr", color_choice: Auto }, network_type: testnet, db_path: "/home/stone/.bitcoin/testnet3/electrs/testnet", daemon_dir: "/home/stone/.bitcoin/testnet3/testnet3", daemon_rpc_addr: V4(127.0.0.1:18332), cookie: Some("thezerg:34983429q980dkdkdsakd"), electrum_rpc_addr: V4(127.0.0.1:60001), monitoring_addr: V4(127.0.0.1:4224), jsonrpc_import: true, index_batch_size: 100, bulk_index_threads: 12, tx_cache_size: 10000, txid_limit: 500, server_banner: "Welcome to electrs (Electrum Rust Server)!" }
DEBUG - Server listening on 127.0.0.1:4224
WARN - failed to connect daemon at 127.0.0.1:18332: Connection refused (os error 111)
DEBUG - Running accept thread
WARN - failed to connect daemon at 127.0.0.1:18332: Connection refused (os error 111)
WARN - failed to connect daemon at 127.0.0.1:18332: Connection refused (os error 111)
.
.
.
```

so it looks like its working from the command line.

**Comment by dagurval** [comment]
> so it looks like its working from the command line.

Weird. Not sure how to reproduce that. Does the qa test run? `./qa/rpc-tests/electrum_basics.py`

File: src/electrum/electrs.cpp (hunk 1 above)
Comment (dagurval) [reply src/electrum/electrs.cpp:31]: Done

File: src/electrum/electrumserver.cpp (hunk 2 above)
Comment (dagurval) [reply src/electrum/electrumserver.cpp]: Fixed

**Comment by dagurval** [comment]
rebased on top of #1646 and #1645

File: Makefile.am [hunk 3]
```diff
@@ -218,6 +218,9 @@ endif
 check-formatting:
 	$(MAKE) -C src $@
 
+electrs: ${PWD}/src/electrs
```
Comment (gandrewstone) [inline Makefile.am]: PWD is not necessarily the out-of-source dir, use top_builddir.  And if ${PWD}/src/electrs does not exist, there's no rule to make it.  Instead use a phony target that points to the real target like this:

```
.PHONY: electrs

electrs: $(top_builddir)/src/electrs

$(top_builddir)/src/electrs:
	$(top_srcdir)/contrib/electrs/build_electrs.py --dst=$(top_builddir)/src

```

**Comment by dagurval** [comment]
Rebased again on top of #1645 and added gitian build support on Linux

File: Makefile.am (hunk 3 above)
Comment (dagurval) [reply Makefile.am]: fixed in 81000427bb55991ecdc35bf4509f12cb287af3ac

**Pull request merged**